src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/XGBoostEvidenceFilter.java

```diff
@@ -0,0 +1,505 @@
+    }
+
+    private CoverageScaledOverlapInfo getClusterOverlapInfo(final BreakpointEvidence evidence) {
+        int clusterNumOverlap = 0;
+        int clusterNumCoherent = 0;
+        int clusterOverlapMappingQuality = 0;
+        int clusterCoherentMappingQuality = 0;
+        double clusterMeanOverlapMappingQuality = 0.0;
+        for (final Iterator<BreakpointEvidence> overlapperItr = evidenceOverlapChecker.overlappers(evidence); overlapperItr.hasNext(); ) {
+            final BreakpointEvidence overlapper = overlapperItr.next();
+            if (overlapper.equals(evidence)) {
+                continue; // don't count self-overlap in cluster features
+            }
+            if(!rawFeatureCache.containsKey(overlapper)) {
+                cacheOverlapInfo(overlapper);
+            }
+            final UnscaledOverlapInfo overlapperFeatureCache = rawFeatureCache.get(overlapper);
+            clusterNumOverlap = Math.max(clusterNumOverlap, overlapperFeatureCache.numOverlap);
+            clusterNumCoherent = Math.max(clusterNumCoherent, overlapperFeatureCache.numCoherent);
+            clusterOverlapMappingQuality = Math.max(clusterOverlapMappingQuality, overlapperFeatureCache.overlapMappingQuality);
+            clusterCoherentMappingQuality = Math.max(clusterCoherentMappingQuality, overlapperFeatureCache.coherentMappingQuality);
+
+            final double meanOverlapMappingQuality = ((double) overlapperFeatureCache.overlapMappingQuality)
+                                                    / overlapperFeatureCache.numOverlap;
+            clusterMeanOverlapMappingQuality = Math.max(clusterMeanOverlapMappingQuality, meanOverlapMappingQuality);
+        }
+
+        return new CoverageScaledOverlapInfo(clusterNumOverlap, clusterNumCoherent, clusterOverlapMappingQuality,
+                clusterCoherentMappingQuality, clusterMeanOverlapMappingQuality, readMetadata.getCoverage());
+    }
+
+    /**
+     * For given BreakpointEvidence, calculate number of overlappers, coherent pieces of BreakpointEvidence, and
+     * the sums of their mapping qualities. Cache this information so that it can be looked up when computing max or
+     * average of these values over the set of overlapping evidence (as in getClusterOverlapInfo).
+     */
+    private void cacheOverlapInfo(final BreakpointEvidence evidence) {
+        int numOverlap = 0;
+        int overlapMappingQuality = 0;
+        int numCoherent = 0;
+        int coherentMappingQuality = 0;
+        for(final EvidenceOverlapChecker.OverlapAndCoherenceIterator overlapperItr
+                = evidenceOverlapChecker.overlappersWithCoherence(evidence);
+            overlapperItr.hasNext();) {
+            final ImmutablePair<BreakpointEvidence, Boolean> itrResults = overlapperItr.next();
+            final BreakpointEvidence overlapper = itrResults.left;
+            if(overlapper.equals(evidence)) {
```

Comment:
On closer inspection (I think I told you to change this to use `equals` instead of `==`?) it looks like `BreakpointEvidence` and subclasses don't have `equals` and `hashCode` methods. Do you want to add them?  Or, I guess we could go back to using `==`, but that makes me nervous about future bugs resulting from serialization or other weirdness.

Reply:
For now I've changed back to ==. Rationale:
    When I made the change I assumed only a horrible person would add a .equals() method to BreakpointEvidence that didn't really ensure equality. That said, I'm really just checking for same object ID, so I'm fine with switching back to ==.
    Thinking about implementing .equals(), I think it might be complicated. I imagine you'd want BreakpointEvidence to check equality of location and evidence type; and ReadEvidence to overload by additionally requiring the same read name?

Reply:
We probably should just implement `equals` methods on all of the `BreakpointEvidence` implementing classes. They should just check equality of all of the fields in each object since that's the expected behavior. Using `==` makes me pretty nervous; it's generally frowned upon for use on Java objects since if the objects get serialized or cloned they no longer look equal to each other even though all the fields match, which is more often than not a bug. Just use the intellij template to generate them.